src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -1059,55 +1065,92 @@ public void StateChanged(long index)
 
         private void NotifyFeaturesAboutStateChange(DatabaseRecord record, long index)
         {
-            lock (_clusterLocker)
+            var taken = false;
+            while (taken == false)
             {
-                DatabaseShutdown.ThrowIfCancellationRequested();
-
-                Debug.Assert(string.Equals(Name, record.DatabaseName, StringComparison.OrdinalIgnoreCase),
-                    $"{Name} != {record.DatabaseName}");
-
-                // index and LastDatabaseRecordIndex could have equal values when we transit from/to passive and want to update the tasks. 
-                if (LastDatabaseRecordIndex > index)
+                Monitor.TryEnter(_clusterLocker, TimeSpan.FromSeconds(5), ref taken);
```

Comment:
Why are you first trying to take the lock, then try to check even if you didn't get it?
Isn't it better to put the `CanSkip` first?

Reply:
We waited for 5 seconds for the lock and got it (or not).. it might be that we don't need to do anything because in that time a newer index already updated the features.

Reply:
Why not put the `CanSkip` before trying to get the lock, as well as after?

Reply:
Done